ramips: add support for Ubiquiti UniFi FlexHD

swegener opened the pull request:
Hardware
--------

- SoC: MediaTek MT7621AT with 128 MiB RAM and 32 MiB Flash
- Wi-Fi: MediaTek MT7603 (b/g/n, 2x2) and MediaTek MT7615 (ac, 4x4)
- Bluetooth: CSR8811 (internal USB, install kmod-bluetooth)

Installation
------------

1. Connect to the booted device at 192.168.1.20 using username/password
   "ubnt".

2. Update the bootloader environment.

   $ fw_setenv devmode TRUE
   $ fw_setenv boot_openwrt "fdt addr \$(fdtcontroladdr);
     fdt rm /signature; bootubnt"
   $ fw_setenv bootcmd "run boot_openwrt"

3. Transfer the OpenWrt sysupgrade image to the device using SCP.

4. Check the mtd partition number for bs / kernel0 / kernel1

   $ cat /proc/mtd

5. Set the bootselect flag to boot from kernel0

   $ dd if=/dev/zero bs=1 count=1 of=/dev/mtdblock4

6. Write the OpenWrt sysupgrade image to both kernel0 as well as kernel1

   $ dd if=openwrt.bin of=/dev/mtdblock6
   $ dd if=openwrt.bin of=/dev/mtdblock7

7. Reboot the device. It should boot into OpenWrt.

Additional Information
----------------------

The LED ring is currently not supported under OpenWrt. The LEDs are not
directly controlled via GPIO pins, but instead are driven by a MCU which
is controlled over I2C. Ubiquiti uses this ledbar concept in other
devices, but I was unable to get it to work with the existing
leds-ubnt-ledbar driver.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
ramips: add support for Ubiquiti UniFi FlexHD

Hardware
--------

- SoC: MediaTek MT7621AT with 128 MiB RAM and 32 MiB Flash
- Wi-Fi: MediaTek MT7603 (b/g/n, 2x2) and MediaTek MT7615 (ac, 4x4)
- Bluetooth: CSR8811 (internal USB, install kmod-bluetooth)

Installation
------------

1. Connect to the booted device at 192.168.1.20 using username/password
   "ubnt".

2. Update the bootloader environment.

   $ fw_setenv devmode TRUE
   $ fw_setenv boot_openwrt "fdt addr \$(fdtcontroladdr);
     fdt rm /signature; bootubnt"
   $ fw_setenv bootcmd "run boot_openwrt"

3. Transfer the OpenWrt sysupgrade image to the device using SCP.

4. Check the mtd partition number for bs / kernel0 / kernel1

   $ cat /proc/mtd

5. Set the bootselect flag to boot from kernel0

   $ dd if=/dev/zero bs=1 count=1 of=/dev/mtdblock4

6. Write the OpenWrt sysupgrade image to both kernel0 as well as kernel1

   $ dd if=openwrt.bin of=/dev/mtdblock6
   $ dd if=openwrt.bin of=/dev/mtdblock7

7. Reboot the device. It should boot into OpenWrt.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

blocktrron commented:
Have you tried modifying the U-Boot environment like explained in the UniFi 6 Lite installation instructions?

https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=fb4d7a9680117a00721936c98ce41eeb2dea95c9

Only try this if you have serial access, as TFTP recovery might not work anymore if done wrong.

The Ubiquiti LEDBar implementation found in the mediatek target relies on the correct enable-gpio being set, were you able to determine this GPIO?

Possible ways to find out are explained here:

https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/mediatek/files-5.4/drivers/leds/leds-ubnt-ledbar.c;h=1f50038b8c4203fb8cf153eb56e9e7451708bb0b;hb=c9137e2ddf97ede1db399dc38acff26feb48ba7e#l14

adschm commented on target/linux/ramips/dts/mt7621_ubnt_unifi-flexhd.dts:
```diff
@@ -0,0 +1,87 @@
+	leds {
+		/delete-node/ dome_blue;
+		/delete-node/ dome_white;
+	};
```
If you need to delete that much from the DTSI, I think it would be better to not use it and just copy the few remaining statements here.

swegener replied:
I have several commits that are not part of this pull request that split the LED definition to separate "ledgpio" and "ledbar" feature .dtsi files, so that each device can include the correct definition.

swegener replied:
See https://github.com/swegener/openwrt/commit/91baae0bdc22084bf36e53af508a9830696f76de for splitting the ledgpio .dtsi to remove the delete-* operations from the flexhd .dts.

adschm replied:
I do not see much point in having a myriad of micro-DTSI files, particularly for LEDs and buttons which have tiny changes of individual GPIO numbers and high/low states for single LED frequently. Please just copy the needed statements from the DTSI and remove the reference as I requested. If you insist, we may discuss the micro-DTSI files in a separate PR, but be aware that I will probably reject it there, too.

swegener replied:
I've merged all required definitions into the device dts.

swegener commented:
> Have you tried modifying the U-Boot environment like explained in the UniFi 6 Lite installation instructions?
> 
> https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=fb4d7a9680117a00721936c98ce41eeb2dea95c9
> 
> Only try this if you have serial access, as TFTP recovery might not work anymore if done wrong.

No, I did not see these instructions and they work for the FlexHD. I've converted the FlexHD image build to FIT and copied over your installation instructions. As the nanoHD uses the same OEM firmware, it can be converted too.

The FlexHD provides a 4 pin header for serial console access, but it is pretty hard to open. The electronics part is screwed in at the top of the tube and seems to be locked with some tabs to prevent it from being unscrewed easily. Also the flat flex connector on the separate ethernet board is easily broken, because of the stress from the angle of the cable.

> The Ubiquiti LEDBar implementation found in the mediatek target relies on the correct enable-gpio being set, were you able to determine this GPIO?
> 
> Possible ways to find out are explained here:
> 
> https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/mediatek/files-5.4/drivers/leds/leds-ubnt-ledbar.c;h=1f50038b8c4203fb8cf153eb56e9e7451708bb0b;hb=c9137e2ddf97ede1db399dc38acff26feb48ba7e#l14

Because of using an older bootloader for my previous installation I was not able to look at the usetled output. And because of the disassembly I'm currently not too fond of opening the device again. OEM FDT switches the uart2, rgmii2 and jtag ports to GPIO mode. jtag is likely switched to prevent access, so it must be one of the pins of uart2 or rgmii2. I'll see what I can come up with.

blocktrron commented:
> As the nanoHD uses the same OEM firmware, it can be converted too.

Things aren't that easy there, as the NanoHD initially shipped with an older bootloader, they've later updated the bootloader to implement FIT as well as signature validation. So while supporting this bootloader is possible (at least until Ubiquiti closes this loophole), it should happen as an additional target in order not to break existing installations. Not great, but this is what we've got.

> Because of using an older bootloader for my previous installation I was not able to look at the usetled output.

Understandable. I can have a look at the bootlaoder binary and see if i can work out the GPIO mapping there. That's what I've did on the U6-LR initially, as Ubiquiti cleverly removed all their proprietary code into their ubntapp stand alone binary :roll_eyes: 

What I've didn't check - You can have a look if the GPIO is assigned on the stock firmware in debugfs `/sys/kernel/debug/gpio`. Also they might not use the i2c bus of the SoC and instead use the i2c-gpio implementation for that matter. This should also be reflected in the stock firmware DTB.

swegener commented:
> > Because of using an older bootloader for my previous installation I was not able to look at the usetled output.
> 
> Understandable. I can have a look at the bootlaoder binary and see if i can work out the GPIO mapping there. That's what I've did on the U6-LR initially, as Ubiquiti cleverly removed all their proprietary code into their ubntapp stand alone binary :roll_eyes: 
> 
> What I've didn't check - You can have a look if the GPIO is assigned on the stock firmware in debugfs `/sys/kernel/debug/gpio`. ~~Also they might not use the i2c bus of the SoC and instead use the i2c-gpio implementation for that matter~~ I've just had a look on that, this is not the case. The Holtek LED controller is on the SoCs I2C bus at 0x30. This should also be reflected in the stock firmware DTB.

Yeah, that what I already figured from decompiling the OEM DTB. The I2C bus is enabled and not switched to GPIO, because the I2C pins are normally used for connecting the GPIO LEDs. There also the 0x30 device definition. It's just the enable GPIO that is missing. IIRC the gpio debugfs only included the reset pin. OEM bootlog mentions resetting the MCU, so there might be multiple pins involved.

blocktrron commented:
Had a look at the ubntapp - Can you try pin 44 / 0x2c for the LEDbar enable-GPIO?

![Screenshot from 2021-05-29 13-13-27](https://user-images.githubusercontent.com/16539439/120068301-18d6a880-c080-11eb-87d5-9e7caa01cb4b.png)

I know this is in the NAND pin group. The OEM firmware might use a different operation mode not requiring toggling the GPIO like the bootloader does. At least that's my theory.

swegener commented:
Sadly I had no success with GPIO44:

```
               enable-gpio = <&gpio 44 GPIO_ACTIVE_HIGH>;
```

```
$ hexdump /sys/firmware/devicetree/base/palmbus@1E000000/i2c@900/led-controller@30/enable-gpio 
0000000 0000 0500 0000 2c00 0000 0000          
000000c
```

```
[    2.575760] i2c-mt7621 1e000900.i2c: clock 400 kHz
[    3.017685] ubnt-ledbar 0-0030: Error initializing LED transaction: 00
[    3.477672] ubnt-ledbar 0-0030: Error initializing LED transaction: 00
[    3.555970] leds blue: Setting an LED's brightness failed (-22)
[    4.007676] ubnt-ledbar 0-0030: Error initializing LED transaction: 00
[    4.085986] leds green: Setting an LED's brightness failed (-22)
[    4.547677] ubnt-ledbar 0-0030: Error initializing LED transaction: 00
[    4.625984] leds red: Setting an LED's brightness failed (-22)
[    4.696087] ubnt-ledbar: probe of 0-0030 failed with error -22
```

I've also tried with GPIO_ACTIVE_LOW, but also no success.

What's the disassembler you used? :) While searching for the pin I've disassembled the gpiodev.ko from the OEM firmware to get insights, but your disassembly looks way nicer to read.

blocktrron commented:
I've used Ghidra to disassemble the binary.

Can you try to increase the amount of milliseconds slept prior to writing to the LED controller (the first one) to 100ms? Assuming the code in above screenshot is in usecs, this might be the issue.

swegener commented:
Increasing the sleep time does not change the behaviour. I've dug deeper into the OEM firmware with Ghidra and GPIO41 is used as a reset line to the MCU. The pin is active low and seems to have a hardware pull-down implemented, i.e. when the GPIO ist not correctly initialized and turned high, the MCU is kept in reset and the LED ring is completely turned off. That was what happended before and now at least the LED ring lights up, but I'm still not getting an answer from the MCU. During GPIO initialization the MCU seems to be reset (the LED ring turns off and on) and I have the suspicion that it needs some kind of initialization sequence. I'll try to extract that from the OEM firmware.

adschm commented on target/linux/ramips/dts/mt7621_ubnt_unifi-flexhd.dts:
```diff
@@ -0,0 +1,140 @@
+	mediatek,mtd-eeprom = <&factory 0x8000>;
+	mtd-mac-address = <&eeprom 0x0>;
+	mtd-mac-address-increment = <2>;
+};
```
No need to have &wlan_Xg split now. Just just merge the nodes and remove the DT labels.

swegener committed:
mediatek: remove gpiod_direction_output()

It's already set to output with GPIOD_OUT_LOW.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: support reading more than one byte of response

There are commands that return more than one byte of response.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: do not use gpiod_set_raw_value()

The polarity of the signal is set in the device dts.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: correctly log i2c response

The read response is in the i2c_response variable. Also use %hhx format,
because we're dealing with a single char.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: cast literal value to char

Or the comparison against a signed char is always true, because the
literal 0xaa is treated as an unsigned int, to which the signed char is
casted during comparison. 0xaa is above the positive values of a signed
char and negative signed char values result in values larger than 0xaa
when casted to unsigned int.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: add support for reset gpio

Some versions of the ledbar MCU have a reset pin. It needs to be
correctly initialized or we might keep the MCU in reset state.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: add led count

The LEDs connected to the MCU are so-called smart LEDs and their signal is
daisy-chained. Because of this, the MCU needs to be told how many LEDs are
connected. It also means the LEDs could be individually controlled, if the MCU
has a command for this.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
mediatek: add initialization after reset

During GPIO initialization the pin state flips and triggers a reset of
the ledbar MCU. It needs to be moved through an initialization sequence
before working correctly.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener commented:
@blocktrron I've managed to get the ledbar working on the FlexHD. That are several changes to the driver necessary, can you take a look over the commits?

Also I think this should be ready for merging.

blocktrron commented on target/linux/mediatek/files-5.10/drivers/leds/leds-ubnt-ledbar.c:
```diff
@@ -67,34 +71,63 @@ static int ubnt_ledbar_apply_state(struct ubnt_ledbar *ledbar)
 	led_msg[UBNT_LEDBAR_TRANSACTION_BLUE_IDX] = ledbar->led_blue.brightness;
 	led_msg[UBNT_LEDBAR_TRANSACTION_GREEN_IDX] = ledbar->led_green.brightness;
 	led_msg[UBNT_LEDBAR_TRANSACTION_RED_IDX] = ledbar->led_red.brightness;
+	led_msg[UBNT_LEDBAR_TRANSACTION_LED_COUNT_IDX] = ledbar->led_count;
```
This byte was zero before, with current core, the `led_count` would always be 1.

swegener replied:
No, the value at `led_msg[UBNT_LEDBAR_TRANSACTION_LED_COUNT_IDX]` was set to 0x01 in the definition of `led_msg` before this commit, defaulting `led_count` to 1 if the dtb entry is missing, just keeps it that way.

blocktrron reviewed (commented):
Awesome. Where did you get the missing information for the LED controller from?

I currently do not have the U6-LR here, will take me ~2 weeks until i have access to it again and verify the LED controller still works.

mans0n commented:
I don't feel duplicating driver files is a good practice... The git log of mediatek target shows codes were moved to generic in such cases.
bcf91fe8849b cc49abc06bca

swegener committed:
generic: move ledbar driver from mediatek target

This moves the ledbar driver to generic, to be also used by the ramips target.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener commented:
> Awesome. Where did you get the missing information for the LED controller from?

It took quite a bit of reading the disassembled code with Ghidra, but was quite fun in the end.

swegener commented:
> I don't feel duplicating driver files is a good practice... The git log of mediatek target shows codes were moved to generic in such cases. [bcf91fe](https://github.com/openwrt/openwrt/commit/bcf91fe8849bf4467637fd5570806837ba93dbd4) [cc49abc](https://github.com/openwrt/openwrt/commit/cc49abc06bca59911caf5b904c3181814636e4da)

I've moved it to generic.

dangowrt commented:
Merged to master. Thanks!
